s2spy/preprocess.py

```diff
@@ -0,0 +1,217 @@
+from typing import Tuple
+from typing import Union
+import numpy as np
+import scipy.stats
+import xarray as xr
+
+
+def _linregress(x: np.ndarray, y: np.ndarray) -> Tuple[float, float]:
+    """Calculate the slope and intercept between two arrays using scipy's linregress.
+
+    Used to make linregress more ufunc-friendly.
+
+    Args:
+        x: First array.
+        y: Second array.
+
+    Returns:
+        slope, intercept
+    """
+    slope, intercept, _, _, _ = scipy.stats.linregress(x, y)
+    return slope, intercept
+
+
+def _trend_linear(data: Union[xr.DataArray, xr.Dataset]) -> dict:
+    """Calculate the linear trend over time.
+
+    Args:
+        data: The input data of which you want to know the trend.
+
+    Returns:
+        Dictionary containing the linear trend information (slope and intercept)
+    """
+    slope, intercept = xr.apply_ufunc(
+        _linregress,
+        data["time"].astype(float),
+        data,
+        input_core_dims=[["time"], ["time"]],
+        output_core_dims=[[], []],
+        vectorize=True,
+    )
+    return {"slope": slope, "intercept": intercept}
+
+
+def _apply_linear_trend(data: Union[xr.DataArray, xr.Dataset], trend: dict):
+    """Apply a previously calclulated linear trend to (new) data."""
+    return data - trend["intercept"] - trend["slope"] * (data["time"].astype(float))
```

Comment:
But what if the applied data does not have the same time axis format? So if the new data has days since 1900 or so, while the trend was calculated on nanoseconds since 1900-01-01 00:00:00 ?

Reply:
Note, if you apply both trend and intercept, you will remove the mean value of the timeseries and therefore not only remove the trend but also center the value around 0 (which I would not want as a user).

Reply:
> center the value around 0 (which I would not want as a user).

Why would you not want that? You will be left with the anomaly. Otherwise you are left with `anomaly + mean`.

To me it feels quite arbitrary to add the mean value at, for example the middle year, back to the data. Why not choose the earliest year, or the latest year? Or always a specific year.

I believe that "detrending" basically always removes the mean by default. The "trend" consists of the "slope" + "intercept".

Reply:
Yes, I agree that the classical statistical definition is like that, but as a scientist manipulating the data this is not what I want.  

We linearly detrend as a proxy for the climate change signal. The intercept is both the mean over all timesteps + the offset due to the trend (e.g., a positive trend coefficient * a positive x-value results in a mean positive offset). In my view, removing the trend-line means something physical, but removing the intercept is more like statistical artifact of the analysis. If you would want to get anomalies, you would need to remove the climatology (which is calculated differently).

Reply:
Just notice this. Yes, I think in meteorology we only want to remove the trend not the intercept. I cannot even think about a usecase for climate study in which the intercept is removed together with the trend. As what @semvijverberg said, if you want to get anomalies (which are centralized around zero), we will remove the climatology. Let's remove the intercept in this function.

Reply:
yes indeed, thanks for the additional info Yang. It is indeed a meteorology/climate science thingy. 

Do note the offset you need to take into account. We also fell into that pitfall in our first attempt. 

I fixed it like this, but there is definitely a more elegant way ;).
```
      # the mean of x-axis * trend is not zero. Determining in-sample offset of trend
      A = np.vstack([x, np.zeros(len(x), dtype="float64")]).T
      trends_offsets = np.dot(A, np.vstack([trends, intercepts])).mean(0)
```